### DISCUSSION TITLE
Consolidate vrfidx and vrf_last_id for vrfs

### DISCUSSION THREAD

**Discussion by jbemmel** [ipspace/netlab#410] [General]
https://github.com/ipspace/netlab/blob/57b215abad87da8a30522ddc7f47ad78c23e6700/netsim/modules/vrf.py#L30

Currently, vrf_last_id is used to generate RD values; it starts from 1
vrfidx is used for l3 evi on srlinux; it starts from 100 ( and srlinux auto-generates the RD from the evi)

While I could manually configure the RD as \<AS\>:1, I think it's better to remove one of these ids

**Comment by ipspace** [comment]
> https://github.com/ipspace/netlab/blob/57b215abad87da8a30522ddc7f47ad78c23e6700/netsim/modules/vrf.py#L30
> 
> Currently, vrf_last_id is used to generate RD values; it starts from 1

We want to have RD values consistent across devices in MPLS/VPN world.

> vrfidx is used for l3 evi on srlinux; it starts from 100 ( and srlinux auto-generates the RD from the evi)

We had that discussion somewhere else. RD doesn't matter (that much) in EVPN world.

> While I could manually configure the RD as <AS>:1

You should not, because that would effectively create global MAC address space which doesn't work that well once you have the same VRRP group in two VRFs.

> I think it's better to remove one of these ids

Well, we do need globally-consistent RD for MPLS/VPN, and if you need an attribute you could use somewhere else, I could add the **id** attribute to VRF so it's usable without reverse-engineering it from RD -- just let me know.

We also need a box-unique VRF ID for things like OSPF process ID, and I have no idea what platform limitations are in that case, so it's safer to use a local counter (because someone could have too many global VRFs, or set VRF ID to some crazy value), thus **vrfidx**.

Which one you want to use in EVPN VRF RD is totally up to you.

**Reply by jbemmel** [reply]
On SR Linux, when left unspecified the RD is auto-generated based on the EVI. What I meant to say, is that I could override that with the RD from the data model - but as you say, it doesn't really matter

What does matter, is that the vrfidx isn't usable as an EVI value for the symmetric L3 VXLAN use case - different nodes may assign different vrfidx for a given vrf. So yes, let's expose a globally unique *id* for each vrf, and then use that as transport_evi

**Reply by ipspace** [reply]
> What does matter, is that the vrfidx isn't usable as an EVI value for the symmetric L3 VXLAN use case

I have no idea what's on behind the scenes of SR Linux, but I haven't found an EVI in the EVPN update messages. They carry RDs, RTs, and VNIs.

> different nodes may assign different vrfidx for a given vrf. So yes, let's expose a globally unique _id_ for each vrf, and then use that as transport_evi

What might transport_evi be? There's transit VNI you have to set in a VRF supporting symmetric IRB.

As promised, I will implement a globally unique VRF _id_, but I still have the feeling you're missing something fundamental (like you finally decided to use netlab-generated RTs/RDs because different vendors use different default values).

**Reply by jbemmel** [reply]
> I have no idea what's on behind the scenes of SR Linux, but I haven't found an EVI in the EVPN update messages. They carry RDs, RTs, and VNIs.
> 

On SR Linux, you need to configure an 'evi' value (range 1..65535) per EVPN enabled vrf (mac-vrf or ip-vrf with symmetric irb).
It only has local significance, BUT it is used in auto-generated RD and RT values - so you may see it on the wire and in interop cases

> > different nodes may assign different vrfidx for a given vrf. So yes, let's expose a globally unique _id_ for each vrf, and then use that as transport_evi
> 
> What might transport_evi be? There's transit VNI you have to set in a VRF supporting symmetric IRB.
> 

What I'm calling 'transport_evi' is the evi value used (locally) in the L3 VNI symmetric IRB case, along with the transport_vni

> As promised, I will implement a globally unique VRF _id_, but I still have the feeling you're missing something fundamental (like you finally decided to use netlab-generated RTs/RDs because different vendors use different default values).

yes I am doing that, because SR Linux defaults to \<AS\>:\<EVI (==vlan) \> and frr uses \<AS\>:\<VNI\> for RTs. I'm just wary of these little details that could cause hard-to-trace issues in large scale setups

**Comment by ipspace** [comment]
Implemented in 6a61761c3165ed4a217ca0aec3a61d3afdf7edeb
